trinity/tx_pool/pool.py

```diff
@@ -0,0 +1,100 @@
+import logging
+from typing import (
+    cast,
+    Iterable,
+    List
+)
+import uuid
+
+from bloom_filter import (
+    BloomFilter
+)
+
+from evm.rlp.transactions import (
+    BaseTransactionFields
+)
+from p2p.eth import (
+    Transactions
+)
+from p2p.peer import (
+    BasePeer,
+    ETHPeer,
+    PeerPool,
+    PeerPoolSubscriber,
+)
+from p2p.service import (
+    BaseService
+)
+
+
+class TxPool(BaseService, PeerPoolSubscriber):
+    """
+    The :class:`~trinity.tx_pool.pool.TxPool` class is responsible for holding and relaying
+    of transactions, represented as :class:`~evm.rlp.transactions.BaseTransaction` among the
+    connected peers.
+
+      .. note::
+
+        This is a minimal vialble implementation that only relays transactions but doesn't actually
+        hold on to them yet. It's still missing many features of a grown up transaction pool.
+    """
+    logger = logging.getLogger("trinity.tx_pool.TxPool")
+
+    def __init__(self, peer_pool: PeerPool) -> None:
+        super().__init__()
+        self._peer_pool = peer_pool
+        # 1m should give us 9000 blocks before that filter becomes less reliable
+        # It should take up about 1mb of memory
+        self._bloom = BloomFilter(max_elements=1000000)
+        self._bloom_salt = str(uuid.uuid4())
+
+    def register_peer(self, peer: BasePeer) -> None:
+        pass
+
+    async def _run(self) -> None:
+        self.logger.info("Running Tx Pool")
+
+        with self.subscribe(self._peer_pool):
+            while True:
+                peer: ETHPeer
+                peer, cmd, msg = await self.wait(
+                    self.msg_queue.get(), token=self.cancel_token)
+
+                if isinstance(cmd, Transactions):
+                    await self._handle_tx(peer, msg)
+
+    async def _handle_tx(self, peer: ETHPeer, txs: List[BaseTransactionFields]) -> None:
+
+        self.logger.debug('Received transactions from %r: %r', peer, txs)
+
+        self._add_txs_to_bloom(peer, txs)
+
+        for receiving_peer in self._peer_pool.peers:
+            receiving_peer = cast(ETHPeer, receiving_peer)
+            if receiving_peer is not peer:
```

Comment:
I've found that in loops like this the following is easier to read.  Thoughts?

```python
for thing in many_things:
    if not some_required_condition(thing):
        continue

    do_for_loop_body_logic(thing)

# instead of
for thing in many_things:
    if some_required_condition(thing):
        do_for_loop_body_logic(thing)
```

Basically, rather than making the context deeper, trying to keep the context as shallow as possible by putting the early exit conditions first and then maintaining the shallow context level for the main function body.

For bodies with a single conditional this isn't a huge improvement, but when we get to having two conditional checks:

```python
for thing in many_things:
    if not some_required_condition(thing):
        continue
    if some_forbidden_condition(thing):
        continue

    do_for_loop_body_logic(thing)

# instead of
for thing in many_things:
    if some_required_condition(thing):
        if not some_forbidden_condition(thing):
            do_for_loop_body_logic(thing)
```

Reply:
I can live with that. I'm personally more used to something like this (python-pseudocode):

```python

things.filter(lambda thing: return thing.is_cool())
          .filter(lambda thing: return thing.rocks())
          .filter(lambda thing: return thing.doesn_it_kitten())
          .for_each(lamda thing: do_stuff(thing))
```

I find this far more readable but I'm not sure if something similarish exists. I guess with lambdas being restricted to single lines in Python this probably doesn't work so well in this ecosystem.

Reply:
Oh and I do agree that rightwards drift is ugly and should be avoided. I'm just not used that much to that loop skipping style.

Reply:
You can accomplish this pretty easily with `toolz/cytoolz` but the result is far less readable.  

```
tuple(pipe(
    things,
    functools.partial(filter, operator.methodcaller('is_cool')),
    functools.partial(filter, operator.methodcaller('rocks')),
    functools.partial(filter, operator.methodcaller('does_it_kitten')),
    functools.partial(map, do_stuff)
))
```

I suspect there is a data pipeline library out there somewhere which allows for a smoother api.

Reply:
>but the result is far less readable

Well...yes...that is.....GOD DAMN UGLY hahaha!

It's fine, it's the ecosystem I signed up for. If that's the price I pay for working on exciting stuff, I'm happy to pay it ;)